thrust/testing/hpx/reduce.cu

```diff
@@ -0,0 +1,138 @@
+#include <thrust/reduce.h>
+
+#include <unittest/unittest.h>
+
+template <typename ExecutionPolicy>
+void TestReduce(ExecutionPolicy exec)
+{
+  using Vector = thrust::device_vector<int>;
+  using T      = typename Vector::value_type;
+
+  Vector v{1, -2, 3};
+
+  // no initializer
+  ASSERT_EQUAL(thrust::reduce(exec, v.begin(), v.end()), 2);
+
+  // with initializer
+  ASSERT_EQUAL(thrust::reduce(exec, v.begin(), v.end(), (T) 10), 12);
+}
+
+void TestReducePar()
+{
+  TestReduce(thrust::hpx::par);
+}
+DECLARE_UNITTEST(TestReducePar);
+
+void TestReduceParOnSequencedExecutor()
+{
+  hpx::execution::sequenced_executor exec{};
+
+  TestReduce(thrust::hpx::par.on(exec));
+}
+DECLARE_UNITTEST(TestReduceParOnSequencedExecutor);
+
+void TestReduceParOnForkJoinExecutor()
+{
+  hpx::execution::experimental::fork_join_executor exec{};
+
+  TestReduce(thrust::hpx::par.on(exec));
+}
+DECLARE_UNITTEST(TestReduceParOnForkJoinExecutor);
+
+void TestReduceParWithAutoChunkSize()
+{
+  hpx::execution::experimental::auto_chunk_size acs{};
+
+  TestReduce(thrust::hpx::par.with(acs));
+}
+DECLARE_UNITTEST(TestReduceParWithAutoChunkSize);
+
+void TestReduceParUnseq()
+{
+  TestReduce(thrust::hpx::par_unseq);
+}
+DECLARE_UNITTEST(TestReduceParUnseq);
+
+void TestReduceParUnseqOnSequencedExecutor()
+{
+  hpx::execution::sequenced_executor exec{};
+
+  TestReduce(thrust::hpx::par_unseq.on(exec));
+}
+DECLARE_UNITTEST(TestReduceParUnseqOnSequencedExecutor);
+
+void TestReduceParUnseqOnForkJoinExecutor()
+{
+  hpx::execution::experimental::fork_join_executor exec{};
+
+  TestReduce(thrust::hpx::par_unseq.on(exec));
+}
+DECLARE_UNITTEST(TestReduceParUnseqOnForkJoinExecutor);
+
+void TestReduceParUnseqWithAutoChunkSize()
+{
+  hpx::execution::experimental::auto_chunk_size acs{};
+
+  TestReduce(thrust::hpx::par_unseq.with(acs));
+}
+DECLARE_UNITTEST(TestReduceParUnseqWithAutoChunkSize);
+
+void TestReduceSeq()
+{
+  TestReduce(thrust::hpx::seq);
+}
+DECLARE_UNITTEST(TestReduceSeq);
+
+void TestReduceSeqOnSequencedExecutor()
+{
+  hpx::execution::sequenced_executor exec{};
+
+  TestReduce(thrust::hpx::seq.on(exec));
+}
+DECLARE_UNITTEST(TestReduceSeqOnSequencedExecutor);
+
+//void TestReduceSeqOnForkJoinExecutor()
+//{
+//  hpx::execution::experimental::fork_join_executor exec{};
+//
+//  TestReduce(thrust::hpx::seq.on(exec));
```

Comment:
is there another executor we could use with `seq`?

Reply:
All existing executors should support the `sync_execute` CP, so they should support being used with `seq`. 

E.g., the [`parallel_executor`](https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/parallel_executor.hpp) should do. The [`sequenced_executor`](https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/sequenced_executor.hpp) is another option.

Reply:
Attempting to rebind `seq` with `fork_join_executor` or `parallel_executor` results in a static assertion failure:

https://github.com/STEllAR-GROUP/hpx/blob/a31463cd0e8049fc09ca5c23632b505c9adb734c/libs/core/execution/include/hpx/execution/executors/rebind_executor.hpp#L71

Reply:
Interesting... The sequenced_executor should work, though. FWIW, a minimal executor supporting `seq` is here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/execution/tests/unit/minimal_sync_executor.cpp#L242-L261

Reply:
The goal is to test changing the executor type, so I think I'll add this minimal executor and use it for all these tests.

Reply:
@hkaiser done, this is now ready.